### PR TITLE
host: Convert string 128-bit UUID to ble_uuid128_t

### PR DESCRIPTION
**Pull request by jrotkiewicz**
Added function that converts 128-bit UUID given as string to a struct ble_uuid128_t.
@sjanc @andrzej-kaczmarek

### CODE REVIEW

File: nimble/host/include/host/ble_uuid.h [hunk 1]
```diff
@@ -168,6 +168,22 @@ void ble_uuid_copy(ble_uuid_any_t *dst, const ble_uuid_t *src);
...
+ *     o "12345678123412341234123456789abc"
+ *
+ * @param str   The source UUID string to convert.
+ * @param uuid  The destination ble_uuid128_t struct.
+ *
+ * @return      0 on success,
+ *              BLE_HS_EINVAL if the specified UUID string has wrong size or
+ *              contains disallowed characters.
+ */
+int ble_str_to_uuid128(const char* str, ble_uuid128_t *uuid);
```
[the remark this answers is not in the document]
Comment (jrotkiewicz) [reply nimble/host/include/host/ble_uuid.h]: Done

File: nimble/host/src/ble_uuid.c [hunk 2]
```diff
@@ -136,6 +186,104 @@ ble_uuid_to_str(const ble_uuid_t *uuid, char *dst)
...
+    uint32_t u16 = 0;
+    uint32_t u32 = 0;
+    uint8_t tmp_rslt = 0;
+
+    int len = (int) strlen(str);
+    str_ptr = &str[len - 2];
+
+    if (len <= BLE_UUID16_STR_MAX_LEN) {
+        uuid->u.type = BLE_UUID_TYPE_16;
+    } else if (len > BLE_UUID16_STR_MAX_LEN &&
```
Comment (sjanc) [inline nimble/host/src/ble_uuid.c:203]: this can be simply  "else if (len <= BLE_UUID32_STR_MAX_LEN)"  cause previous check already handled <= BLE_UUID16_STR_MAX_LEN

will make code a bit easier to follow

File: nimble/host/src/ble_uuid.c [hunk 3]
```diff
@@ -136,6 +186,104 @@ ble_uuid_to_str(const ble_uuid_t *uuid, char *dst)
...
+        }
+
+        if (memcmp(ble_uuid_base, uuid->u128.value, 12) == 0) {
+            uint8_t *tmp_ptr = &uuid->u128.value[12];
+            uint32_t tmp_val32 = 0;
+            for (int i = 0; i < 4; i++) {
+                tmp_val32 |= ((uint32_t)(*tmp_ptr++) << 8 * i);
+            }
+
+            if (tmp_val32 < UINT16_MAX) {
```
Comment (sjanc) [inline nimble/host/src/ble_uuid.c:243]: shouldn't this be <= ?

(could also add test for edge values :-)

File: nimble/host/include/host/ble_uuid.h [hunk 4]
```diff
@@ -253,6 +253,26 @@ void ble_uuid_copy(ble_uuid_any_t *dst, const ble_uuid_t *src);
  */
 char *ble_uuid_to_str(const ble_uuid_t *uuid, char *dst);
 
+/**
+ * @brief Converts the specified UUID string to ble_uuid_any_t representation.
+ *        If the UUID is recognised as Bluetooth SIG UUID, it will provide its
+ *        32-bit or 16-bit representation.
+ *
+ * Example 128-bit string representations:
+ *     o "12345678-1234-1234-1234-123456789abc\0"
```
Comment (sjanc) [inline nimble/host/include/host/ble_uuid.h:262]: string is by definition null terminated

File: nimble/host/src/ble_uuid.c [hunk 5]
```diff
@@ -136,6 +186,104 @@ ble_uuid_to_str(const ble_uuid_t *uuid, char *dst)
...
+    uint8_t *u8p;
+    const char *str_ptr;
+    uint32_t u16 = 0;
+    uint32_t u32 = 0;
+    uint8_t tmp_rslt = 0;
+
+    int len = (int) strlen(str);
+    str_ptr = &str[len - 2];
+
+    if (len <= BLE_UUID16_STR_MAX_LEN) {
```
Comment (sjanc) [inline nimble/host/src/ble_uuid.c:201]: should this also validate min len, and maybe that len is even?

if I pass "1" it shouldn't probably start parsing (and str_ptr will point before str buffer)

File: nimble/host/src/ble_uuid.c (hunk 2 above)
Comment (jrotkiewicz) [reply nimble/host/src/ble_uuid.c:203]: Done

File: nimble/host/src/ble_uuid.c (hunk 3 above)
Comment (jrotkiewicz) [reply nimble/host/src/ble_uuid.c:243]: Yes, added tests ; )

File: nimble/host/include/host/ble_uuid.h (hunk 4 above)
Comment (jrotkiewicz) [reply nimble/host/include/host/ble_uuid.h:262]: oops :D fixed

File: nimble/host/src/ble_uuid.c (hunk 5 above)
Comment (jrotkiewicz) [reply nimble/host/src/ble_uuid.c:201]: Added check and test

**Review by sjanc — APPROVED** [review]
